Allow named pipes / FIFOs when using `FileIO` (fixes #25328).

nathankleyn opened the pull request:
`FileIO` was using NIO's `isRegularFile` to check whether the path
passed was a file and not a directory (added in #21801) - however this
has the side-effect of also excluding things that are not considered
regular files in POSIX terminology, like named pipes / FIFOs.

The requirement has been swapped to only check if something is not a
directory, allowing anything else - this could mean a user can try to
open other non-regular files like devices, but that check is probably
not necessary.

Tests have been added to ensure that named pipes / FIFOs can be read,
but the POSIX standard `mkfifo` is the only way to create one (the JVM
does not supply a cross-platform method for doing that) - this should
still work on Windows thanks to the Linux Subsystem support for this
command.

nathankleyn committed:
Allow named pipes / FIFOs when using `FileIO` (fixes #25328).

`FileIO` was using NIO's `isRegularFile` to check whether the path
passed was a file and not a directory (added in #21801) - however this
has the side-effect of also excluding things that are not considered
regular files in POSIX terminology, like named pipes / FIFOs.

The requirement has been swapped to only check if something is not a
directory, allowing anything else - this could mean a user can try to
open other non-regular files like devices, but that check is probably
not necessary.

lightbend-cla-validator commented:
Hi @nathankleyn,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

nathankleyn commented:
One of the tests in here is having to use the shell to create the FIFO to ensure it works (there's no provided way in Java to create one AFAIK). Interested in feedback about the approach in regards to that, as support on Windows would only work via Windows' relatively new Linux Subsystem.

giftig commented on akka-stream-tests/src/test/scala/akka/stream/io/FileSourceSpec.scala:
```diff
@@ -59,6 +59,20 @@ class FileSourceSpec extends StreamSpec(UnboundedMailboxConfig) {
   }
 
+  val directoryInsteadOfFile = {
+    val f = Files.createTempDirectory(fs.getPath("/"), "directory-instead-of-file")
```
Unnecessary inner val? You're just returning it immediately anyway.

nathankleyn replied:
Sorry, that's daft - will just return.

giftig commented on akka-stream-tests/src/test/scala/akka/stream/io/FileSourceSpec.scala:
```diff
@@ -59,6 +59,20 @@ class FileSourceSpec extends StreamSpec(UnboundedMailboxConfig) {
+    val d = Files.createTempDirectory(fs.getPath("/"), "file-source-spec")
+    val f = d.resolve("test-fifo")
+    // Java or NIO does not supply a way to create named pipes/FIFOs, so we use the POSIX command `mkfifo`.
+    Runtime.getRuntime.exec(s"mkfifo $f")
```
Does this actually work? `fs` is an in-memory filesystem created with `JimFs`; it looks like this is just going to create a fifo on the actual filesystem (and it probably won't have permission, nor be portable)

nathankleyn replied:
Oops, I did not realise `fs` was special - and that is probably the nail in the coffin for this test. [I was already concerned with the portability of this test on at least Windows](https://github.com/akka/akka/pull/25333#issuecomment-403783216) - so I'm wondering if we should even have a test for FIFOs specifically?

giftig replied:
Yeah I was about to comment on portability but then I checked the source file to see what `fs` was (because I noticed it using `/` and it seemed very unlikely to me that the akka test suite would try to create things in the root filesystem) and discovered JimFs.

I had a look at JimFs and didn't see any mention of fifos so it doesn't look very straightforward to test in a portable way. It'd be a shame to not be able to test this, though, as it means there's no guarantee your fifo will continue working so your issue may recur.

giftig replied:
Unfortunately it doesn't look like `FileSource` is structured in a way which would make injecting a mock `File` into it very easy either, though, and such a test would stand out as being quite different to the others.

Maybe someone on the akka team has a better suggestion.

giftig commented on akka-stream-tests/src/test/scala/akka/stream/io/FileSourceSpec.scala:
```diff
@@ -107,6 +121,17 @@ class FileSourceSpec extends StreamSpec(UnboundedMailboxConfig) {
+      p.request(1)
+      p.expectNext().utf8String should ===(TestText.splitAt(chunkSize)._1)
+      p.cancel()
+      Await.result(future, 3.seconds) === createSuccessful(chunkSize)
```
The spec class extends `ScalaFutures` (from AkkaSpec, via StreamSpec) so you can just use `.futureValue`. I see other tests are using `Await.result` though; I wonder if there's a reason for using that over `futureValue`? Among other things, `futureValue` will allow scaling with scalatest's patience config etc.

nathankleyn replied:
I think if the above stands, this particular test is likely not going to be there, but can certainly update the other tests to use `futureValue` and am happy to do so!

giftig commented on akka-stream-tests/src/test/scala/akka/stream/io/FileSourceSpec.scala:
```diff
@@ -203,6 +228,19 @@ class FileSourceSpec extends StreamSpec(UnboundedMailboxConfig) {
+      c.expectSubscription()
+
+      val error = c.expectError()
+      error shouldBe a[IllegalArgumentException]
```
If you want to be grammatically correct you should use `an[IllegalArgumentException]` :D

nathankleyn replied:
Doh. 😂

nathankleyn commented:
@giftig I've pushed the two easy fixes -  but the other one stands on a choice as to whether it's possible/advisable to even test using FIFOs at all given the challenges with portability and `JimFs` not supporting anything of that ilk.

Let me know what you think is best and I can update this branch ASAP. 👍

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/io/FileSourceSpec.scala:
```diff
@@ -59,6 +59,17 @@ class FileSourceSpec extends StreamSpec(UnboundedMailboxConfig) {
+    Runtime.getRuntime.exec(s"mkfifo $f")
+    Files.newBufferedWriter(f, UTF_8).append(TestText).close()
+    f
+  }
```
Let's just skip testing this, forking shell command processes from a test does feel like it makes for fragile tests and a high chance of leaking stuff over time etc.

The reason we use jimfs is to make the test work consistently and fast across machines.

johanandren commented on akka-stream-tests/src/test/scala/akka/stream/io/FileSourceSpec.scala:
```diff
@@ -203,6 +225,19 @@ class FileSourceSpec extends StreamSpec(UnboundedMailboxConfig) {
+      error shouldBe an[IllegalArgumentException]
+
+      Await.result(r, 3.seconds.dilated).status.isFailure shouldBe true
+    }
```
Prefer `futureValue` where possible. It will pick up a reasonable default value. Awaits here are probably just a leftover before Scalatest had `futureValue`

johanandren reviewed (commented):
Looks like the fix is good, but the added some comments for the test

nathankleyn commented:
Many thanks for reviewing @johanandren! I've removed the test in question, and cleaned up all usages of `Await` in that one spec in favour of `.futureValue`. I've rebased on the latest `master` as well, so should be good to re-review whenever you're free. 🤞

johanandren reviewed (approved):
LGTM, thanks!

giftig reviewed (approved):
LGTM

giftig commented:
@johanandren /echo "OK TO TEST"? :D

raboof commented:
OK TO TEST

johanandren commented:
Fixes #25328
